extensions/webirc.md

```diff
@@ -0,0 +1,60 @@
+---
+title: IRCv3 WebIRC extension
+layout: spec
+work-in-progress: true
+copyrights:
+  -
+    name: "Evan Magaliff"
+    period: "2017"
+    email: "muffinmedic@kiwiirc.com"
+  -
+    name: "Darren Whitlen"
+    period: "2017"
+    email: "prawnsalad@kiwiirc.com"
+---
+## Purpose
+This specification is intended to document the existing `WEBIRC` command that is already in use by multiple IRC servers and WebIRC gateways to provide a standard specification for implementation. This specification does not add any new features or changes, although new features, such as use of public-key cryptography instead of password authentication, can be submitted as extensions.
+
+## Description
+When a user connects through an indirect connection to the IRC server, the user's actual IP address and hostname are not visible. The `WEBIRC` command allows the WebIRC gateway to pass on the user's actual IP address and hostname to the IRC server. This information can be used by the network as if the user was connecting directly.
+
+## Format
+The `WEBIRC` command MUST be the first command sent from the WebIRC gateway to the IRC server and MUST be sent before capability negotiation.
+
+The `WEBIRC` command takes four parameters: `password `, `user`, `hostname`, and `ip`.
+- `password` Password authenticating the WebIRC gateway to the IRC server which must be agreed upon ahead of time
+- `user` WebIRC gateway service name
+- `hostname` Resolved hostname of the user's IP address
+- `ip` IPv4 or IPv6 address of the connecting user. If the IP address cannot be resolved to a DNS name, the IP address MUST be sent as both the `hostname` and `ip`.
+
+If forward and reverse DNS do not match, the IP address SHOULD be sent as the `hostname` and the unverified `hostname` SHOULD NOT be sent (see [Security Considerations](#security-considerations)). If the connection fails to the IRC server, the WebIRC gateway MUST return an `ERROR` with an appropriate message and terminate the connection.
+
+### Examples
+Generic format.
+
+    WEBIRC password user hostname ip
```

Comment:
I think overall (if this is meant to help new implementers out as well) it would help to clarify which component is being asked (that of the server or of the user), as it's rather confusing and unclear to me.

Take the following for instance:
```
- `user` WebIRC gateway service name
   ^^^^
- `hostname` Resolved hostname of the user's IP address
                                      ^^^^
```

It doesn't help that `user` is also an irc verb for "username", or "ident". it's also called gatewayname a bit lower. It's confusing when it's mixed together like this, and I would clarify by prefixing something like `server_${param}` and `user_${param}`.